MEMMODEL_CONSTRAINTS.md

```diff
@@ -103,6 +103,19 @@ may perceive in that statement.)
   for memory used for atomics.  The situation is somewhat better in
   well-typed C++ but reality is complicated.
 
+## Implementation targeting C and C++ or LLVM
+
+* _Targeting C or C++_: it should be possible to implement
+  SharedArrayBuffer using C or C++.  The C/C++11 memory model allows
+  accesses with a number of orderings, and there should be a mapping
+  from SharedArrayBuffer accesses to C/C++11 orderings. Racy
+  SharedArrayBuffer accesses are required to produce defined values,
+  so cannot be implemented using C/C++ non-atomics.
+* _Targeting LLVM_: implementations may also decide to target
+  LLVM, which may provide for more optimization than targeting C/C++
+  and compiling to LLVM. In particular, LLVM allows unordered access,
+  which does not require per-variable sequential consistency.
```

Comment:
> I'm not sure all C/C++ users will use the locking primitives, some programs may implement lock-free data structures using atomic accesses, which would then compile to SAB accesses.

Lock-free data structures use `std::atomic`. Others use `std::mutex`. This all maps to SAB. I'm not sure I understand what you're disagreeing with.

> I'm also not sure we'll want to ignore LLVM unordered, we could map SAB non-atomic to LLVM unordered, depending on whether SAB non-atomics are required to be coherent.

It's premature to consider `unordered` since SAB doesn't support anything but `seq_cst`.

> AFAIK the Rust memory model is built on top of the LLVM model (with extra guarantees for safe code).

Would be good to get @pcwalton's in put on this.